R/dataframe.R

```diff
@@ -157,6 +160,9 @@ as_data_frame.data.frame <- function(x, ...) {
 #' @export
 #' @rdname as_data_frame
 as_data_frame.list <- function(x, validate = TRUE, ...) {
+
+  x <- compact(x)
```

Comment:
The implementation of compact() filters NULL values. I am missing the connection with the invariant you stated (which should be documented, I agree).